compiler/ilgen/VirtualMachineRegister.hpp

```diff
@@ -0,0 +1,179 @@
+      _pointerToRegisterType(pointerToRegisterType),
+      _elementType(pointerToRegisterType->baseType()),
+      _adjustByStep(adjustByStep)
+      {
+      Reload(b);
+      }
+
+  
+   /**
+    * @brief write the simulated register value to the virtual machine
+    * @param b the builder where the operation will be placed to store the virtual machine value
+    */
+   virtual void Commit(TR::IlBuilder *b)
+      {
+      b->StoreAt(
+            _addressOfRegister,
+      b->   Load(_localName));
+      }
+
+   /**
+    * @brief transfer the current virtual machine register value to the simulated local variable
+    * @param b the builder where the operation will be placed to load the virtual machine value
+    */
+   virtual void Reload(TR::IlBuilder *b)
+      {
+      b->Store(_localName,
+      b->   LoadAt((TR::IlType *)_pointerToRegisterType,
+               _addressOfRegister));
+      }
+
+   /**
+    * @brief used in the compiled method to load the (simulated) register's value
+    * @param b the builder where the operation will be placed to load the simulated value
+    * @returns TR:IlValue * for the loaded simulated register value
+    */
+   TR::IlValue *Load(TR::IlBuilder *b)
+      {
+      return b->Load(_localName);
+      }
+
+   /**
+    * @brief used in the compiled method to store to the (simulated) register's value
+    * @param b the builder where the operation will be placed to store to the simulated value
+    */
+   void Store(TR::IlBuilder *b, TR::IlValue *value)
+      {
+      b->Store(_localName, value);
+      }
+
+   /**
+    * @brief used in the compiled method to add to the (simulated) register's value
+    * @param b the builder where the operation will be placed to add to the simulated value
+    * @param amount the TR::IlValue that should be added to the simulated value, after multiplication by _adjustByStep
+    * Adjust() is really a convenience function for the common operation of adding a value to the register. More
+    * complicated operations (e.g. multiplying the value) can be built using Load() and Store() if needed.
+    */
+   virtual void Adjust(TR::IlBuilder *b, TR::IlValue *amount)
+      {
+      b->Store(_localName,
+      b->   Add(
+      b->      Load(_localName),
+      b->      Mul(
+                  amount,
+      b->         ConstInteger(_elementType, _adjustByStep))));
+      }
+
+   /**
+    * @brief used in the compiled method to add to the (simulated) register's value
+    * @param b the builder where the operation will be placed to add to the simulated value
+    * @param amount the constant value that should be added to the simulated value, after multiplication by _adjustByStep
+    * Adjust() is really a convenience function for the common operation of adding a value to the register. More
+    * complicated operations (e.g. multiplying the value) can be built using Load() and Store() if needed.
+    */
+   virtual void Adjust(TR::IlBuilder *b, int64_t amount)
```

Comment:
Maybe this can call the Adjust routine taking an arg of type TR::IlValue * to keep the (small) logic in one place. 

Wondered about the name "Adjust" which is'nt the most descriptive, but it is an aesthetic point.